src/Http/Http/src/ReadOnlyPipeStream.cs

```diff
@@ -0,0 +1,223 @@
+    public class ReadOnlyPipeStream : Stream
+    {
+        private readonly PipeReader _pipeReader;
+
+        /// <summary>
+        /// Creates a new ReadOnlyPipeStream
+        /// </summary>
+        /// <param name="pipeReader"></param>
+        public ReadOnlyPipeStream(PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader;
+        }
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanRead => true;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override int WriteTimeout
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void Flush()
+        {
+        }
+
+        /// <inheritdoc />
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+        }
+
+        /// <inheritdoc />
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            var task = ReadAsync(buffer, offset, count, default(CancellationToken), state);
+            if (callback != null)
+            {
+                task.ContinueWith(t => callback.Invoke(t));
+            }
+            return task;
+        }
+
+        /// <inheritdoc />
+        public override int EndRead(IAsyncResult asyncResult)
+        {
+            return ((Task<int>)asyncResult).GetAwaiter().GetResult();
+        }
+
+        private Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken, object state)
+        {
```

Comment:
Most of this is copied from: https://github.com/aspnet/KestrelHttpServer/blob/master/src/Kestrel.Core/Internal/Http/HttpRequestStream.cs

Reply:
If `ReadOnlyPipeStream` and `WriteOnlyPipeStream` are in the `System.IO.Pipelines` namespace, and not specific to this use-case (it appears to be generic), would it not be better to include this in the System.IO.Pipelines package so it can be used elsewhere?
I know there is a package on NuGet called [System.IO.Pipelines.Extensions](https://www.nuget.org/packages/System.IO.Pipelines.Extensions) which claims to be "Stream pipeline adapters implementation"
I know that I would be very happy if this had an official implementation available outside of aspnetcore as well.

Reply:
The plan is to eventually port this to System.IO.Pipelines in dotnet/corefx. This is temporary so we can dogfood it in aspnetcore.

Reply:
@CZEMacLeod System.IO.Pipelines.Extensions is old and probably busted by now.